### PR TITLE
Feature//Implement Remote Data Layer

### PR DESCRIPTION
**Pull request by Prangula**

### CODE REVIEW

File: .idea/deploymentTargetSelector.xml
Comment (etozheluka) [inline .idea/deploymentTargetSelector.xml:1]: add .idea to gitignore

File: app/src/main/java/com/imdbmovieapp/data/remote/dto/detailMovieDto/DetailMovieDto.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    val release_date: String,
+    val revenue: Int,
+    val runtime: Int,
+    val spoken_languages: List<SpokenLanguage>,
+    val status: String,
+    val tagline: String,
+    val title: String,
+    val video: Boolean,
+    val vote_average: Double,
+    val vote_count: Int
```
Comment (etozheluka) [inline app/src/main/java/com/imdbmovieapp/data/remote/dto/detailMovieDto/DetailMovieDto.kt]: you can remove properties which you dont need

File: app/src/main/java/com/imdbmovieapp/data/remote/api/MoviesApi.kt [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import com.imdbmovieapp.data.remote.dto.searchMoviesDto.SearchMoviesDto
+import com.imdbmovieapp.data.remote.dto.topRatedMoviesDto.TopRatedMoviesDto
+import retrofit2.http.GET
+import retrofit2.http.Path
+import retrofit2.http.Query
+
+interface MoviesApi {
+
+    @GET("/3/movie/popular")
+    suspend fun getPopularMovies(): List<PopularMoviesDto>
```
Comment (etozheluka) [inline app/src/main/java/com/imdbmovieapp/data/remote/api/MoviesApi.kt]: return Response<List<PopularMoviesDto>>

File: app/src/main/java/com/imdbmovieapp/data/remote/api/MoviesApi.kt [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+    @GET("/3/movie/top_rated")
+    suspend fun getTopRatedMovies(): List<TopRatedMoviesDto>
```
Comment (etozheluka) [inline app/src/main/java/com/imdbmovieapp/data/remote/api/MoviesApi.kt]: return Response<List< TopRatedMoviesDto >>

File: app/src/main/java/com/imdbmovieapp/data/remote/api/MoviesApi.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+    @GET("/3/movie/{movie_id}")
+    suspend fun getDetailMovie(@Path("movie_id") movieId: Int): DetailMovieDto
+
+    @GET("/3/search/movie")
+    suspend fun getSearchMovies(@Query("query") query: String): List<SearchMoviesDto>
+
+    @GET("/3/genre/movie/list")
+    suspend fun getGenres(): List<GenreMovieDto>
```
Comment (etozheluka) [inline app/src/main/java/com/imdbmovieapp/data/remote/api/MoviesApi.kt]: response gives you better error handling, also you can access to headers and etc

File: app/src/main/java/com/imdbmovieapp/data/remote/dto/detailMovieDto/Genre.kt [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package com.imdbmovieapp.data.remote.dto.detailMovieDto
+
+data class Genre(
```
Comment (etozheluka) [inline app/src/main/java/com/imdbmovieapp/data/remote/dto/detailMovieDto/Genre.kt]: MovieGenreDto

File: app/src/main/java/com/imdbmovieapp/data/remote/dto/detailMovieDto/ProductionCompany.kt [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package com.imdbmovieapp.data.remote.dto.detailMovieDto
+
+data class ProductionCompany(
```
Comment (etozheluka) [inline app/src/main/java/com/imdbmovieapp/data/remote/dto/detailMovieDto/ProductionCompany.kt]: MovieProductionCompanyDto

File: app/src/main/java/com/imdbmovieapp/data/remote/dto/detailMovieDto/ProductionCountry.kt [hunk 7]
```diff
@@ -0,0 +1,6 @@
+package com.imdbmovieapp.data.remote.dto.detailMovieDto
+
+data class ProductionCountry(
```
Comment (etozheluka) [inline app/src/main/java/com/imdbmovieapp/data/remote/dto/detailMovieDto/ProductionCountry.kt]: MovieProductionCountryDto

File: app/src/main/java/com/imdbmovieapp/data/remote/dto/detailMovieDto/SpokenLanguage.kt [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package com.imdbmovieapp.data.remote.dto.detailMovieDto
+
+data class SpokenLanguage(
+    val english_name: String,
+    val iso_639_1: String,
+    val name: String
```
Comment (etozheluka) [inline app/src/main/java/com/imdbmovieapp/data/remote/dto/detailMovieDto/SpokenLanguage.kt]: check other data classes, do you really need them ?

File: app/src/main/java/com/imdbmovieapp/data/remote/dto/genreMovieDto/Genre.kt [hunk 9]
```diff
@@ -0,0 +1,6 @@
+package com.imdbmovieapp.data.remote.dto.genreMovieDto
+
+data class Genre(
+    val id: Int,
+    val name: String
+)
```
Comment (etozheluka) [inline app/src/main/java/com/imdbmovieapp/data/remote/dto/genreMovieDto/Genre.kt]: you have tho Genre data classes

File: app/src/main/java/com/imdbmovieapp/data/remote/dto/popularMoviesDto/Result.kt [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+    val original_language: String,
+    val original_title: String,
+    val overview: String,
+    val popularity: Double,
+    val poster_path: String,
+    val release_date: String,
+    val title: String,
+    val video: Boolean,
+    val vote_average: Double,
+    val vote_count: Int
```
Comment (etozheluka) [inline app/src/main/java/com/imdbmovieapp/data/remote/dto/popularMoviesDto/Result.kt]: you have 2 result dto classes

**Review by etozheluka — CHANGES_REQUESTED** [review]
please check all the parametrs and refactor your dtos.

File: app/src/main/java/com/imdbmovieapp/data/remote/dto/detailMovieDto/MovieDetailDto.kt [hunk 11]
```diff
@@ -0,0 +1,13 @@
+package com.imdbmovieapp.data.remote.dto.detailMovieDto
+
+import com.imdbmovieapp.data.remote.dto.genreMovieDto.Genre
+
+data class MovieDetailDto(
+    val genres: List<Genre>,
+    val id: Int,
```
Comment (Davit-Tsiskarashvili) [inline app/src/main/java/com/imdbmovieapp/data/remote/dto/detailMovieDto/MovieDetailDto.kt:7]: generally it's good practice to start with id

File: app/src/main/java/com/imdbmovieapp/data/remote/dto/detailMovieDto/MovieDetailDto.kt [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+    val overview: String,
+    val poster_path: String,
+    val release_date: String,
+    val title: String,
+    val vote_average: Double,
```
Comment (Davit-Tsiskarashvili) [inline app/src/main/java/com/imdbmovieapp/data/remote/dto/detailMovieDto/MovieDetailDto.kt]: Try not to use commas for the last parameters. You have it in MoviesResult as well. It's a matter of style, but when you will be in Space project, you will have to fix bot comments about it :D

File: app/src/main/java/com/imdbmovieapp/data/remote/dto/popularMoviesDto/MoviesResult.kt [hunk 13]
```diff
@@ -0,0 +1,9 @@
+package com.imdbmovieapp.data.remote.dto.popularMoviesDto
+
+data class MoviesResult(
+    val genre_ids: List<Int>,
```
Comment (Davit-Tsiskarashvili) [inline app/src/main/java/com/imdbmovieapp/data/remote/dto/popularMoviesDto/MoviesResult.kt]: id first and comma on the last property

File: app/src/main/java/com/imdbmovieapp/data/remote/dto/searchMoviesDto/SearchMoviesDto.kt [hunk 14]
```diff
@@ -0,0 +1,8 @@
+package com.imdbmovieapp.data.remote.dto.searchMoviesDto
+
+data class SearchMoviesDto(
+    val page: Int,
+    val results: List<Any>,
+    val total_pages: Int,
+    val total_results: Int
```
Comment (Davit-Tsiskarashvili) [inline app/src/main/java/com/imdbmovieapp/data/remote/dto/searchMoviesDto/SearchMoviesDto.kt]: We don't name variables with "_". Name it like totalPages, totalResults (fix names in other classes as well)

File: .idea/gradle.xml
Comment (bchmsl) [inline .idea/gradle.xml:1]: Same as I wrote in previous PR:
https://github.com/Prangula/ImdbMovieApp/pull/2
Add `.idea` files to `gitignore`

Simply add 
`.idea/`
to gitignore file and for already tracked .idea files you can run following command
`git rm -r --cached .idea`
and commit changes to your branch
`git commit -m "Add .idea to .gitignore or smth, some commit message"`

File: app/src/main/java/com/imdbmovieapp/data/remote/dto/detailMovieDto/MovieDetailDto.kt [hunk 15]
```diff
@@ -0,0 +1,13 @@
+package com.imdbmovieapp.data.remote.dto.detailMovieDto
+
+import com.imdbmovieapp.data.remote.dto.genreMovieDto.Genre
+
+data class MovieDetailDto(
+    val genres: List<Genre>,
+    val id: Int,
+    val overview: String,
+    val poster_path: String,
+    val release_date: String,
```
Comment (bchmsl) [inline app/src/main/java/com/imdbmovieapp/data/remote/dto/detailMovieDto/MovieDetailDto.kt]: Use `@SerializedName` for Gson and `@Json(name="poster_path")` for Moshi annotations for underscore parameter names and use camelCase approach in every Kotlin file.

File: app/src/main/java/com/imdbmovieapp/data/remote/dto/popularMoviesDto/PopularMoviesDto.kt [hunk 16]
```diff
@@ -0,0 +1,8 @@
+package com.imdbmovieapp.data.remote.dto.popularMoviesDto
+
+data class PopularMoviesDto(
+    val page: Int,
+    val moviesResults: List<MoviesResult>,
+    val total_pages: Int,
+    val total_results: Int
```
Comment (bchmsl) [inline app/src/main/java/com/imdbmovieapp/data/remote/dto/popularMoviesDto/PopularMoviesDto.kt]: Try to sort properties by their importance (for example, in my opinion, `movieResults` property is more important than `page`) and move it above or below others

File: app/src/main/java/com/imdbmovieapp/data/remote/dto/topRatedMoviesDto/TopRatedMoviesDto.kt [hunk 17]
```diff
@@ -0,0 +1,10 @@
+package com.imdbmovieapp.data.remote.dto.topRatedMoviesDto
+
+import com.imdbmovieapp.data.remote.dto.popularMoviesDto.MoviesResult
+
+data class TopRatedMoviesDto(
+    val page: Int,
+    val results: List<MoviesResult>,
+    val total_pages: Int,
+    val total_results: Int
```
Comment (bchmsl) [inline app/src/main/java/com/imdbmovieapp/data/remote/dto/topRatedMoviesDto/TopRatedMoviesDto.kt]: Same about important properties

File: app/src/main/java/com/imdbmovieapp/data/remote/dto/popularMoviesDto/MoviesResult.kt [hunk 18]
```diff
@@ -0,0 +1,9 @@
+package com.imdbmovieapp.data.remote.dto.popularMoviesDto
+
+data class MoviesResult(
```
Comment (bchmsl) [inline app/src/main/java/com/imdbmovieapp/data/remote/dto/popularMoviesDto/MoviesResult.kt]: Add DTO/Dto suffix for all separate Dtos (which are not inside another class)

**Comment by bchmsl** [comment]
About styling and cleaning your PR, same applies to this PR as I wrote on the previous one
https://github.com/Prangula/ImdbMovieApp/pull/2#issuecomment-2358869966

File: app/src/main/java/com/imdbmovieapp/data/remote/dto/popularMoviesDto/MoviesResultDto.kt [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+    @Json(name = "genre_ids")
+    val genreIds: List<Int>,
+    @SerializedName("poster_path")
+    @Json(name = "poster_path")
+    val posterPath: String,
+    @SerializedName("release_date")
+    @Json(name = "release_date")
+    val releaseDate: String,
+    val title: String
+)
```
Comment (Davit-Tsiskarashvili) [inline app/src/main/java/com/imdbmovieapp/data/remote/dto/popularMoviesDto/MoviesResultDto.kt]: This is not correct. As Bachana wrote in his comment, you should use @SerializedName if you are using Gson library, but if you are using Moshi library, you should use @Json

**Review by Davit-Tsiskarashvili — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/imdbmovieapp/data/remote/dto/popularMoviesDto/MoviesResultDto.kt (hunk 19 above)
Comment (Davit-Tsiskarashvili) [reply app/src/main/java/com/imdbmovieapp/data/remote/dto/popularMoviesDto/MoviesResultDto.kt]: As I've checked your Gradle, you are using Gson. Therefore remove @Json(name = " ") from every data class

Comment (Prangula) [reply app/src/main/java/com/imdbmovieapp/data/remote/dto/popularMoviesDto/MoviesResultDto.kt]: ok, understand now

Comment (Prangula) [reply app/src/main/java/com/imdbmovieapp/data/remote/dto/popularMoviesDto/MoviesResultDto.kt]: so i just have to leave for example:  @SerializedName("poster_path") and remove @Json(name = "poster_path")

Comment (Davit-Tsiskarashvili) [reply app/src/main/java/com/imdbmovieapp/data/remote/dto/popularMoviesDto/MoviesResultDto.kt]: yes

**Review by Davit-Tsiskarashvili — APPROVED** [review]

File: app/src/main/java/com/imdbmovieapp/data/remote/api/MoviesApi.kt [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+    suspend fun getTopRatedMovies(): Response<List<TopRatedMoviesDto>>
+
+    @GET("/3/movie/{movie_id}")
+    suspend fun getDetailMovie(@Path("movie_id") movieId: Int): Response<MovieDetailDto>
+
+    @GET("/3/search/movie")
+    suspend fun getSearchMovies(@Query("query") query: String): Response<List<SearchMoviesDto>>
+
+    @GET("/3/genre/movie/list")
+    suspend fun getGenres(): Response<List<MovieGenreDto>>
```
Comment (etozheluka) [inline app/src/main/java/com/imdbmovieapp/data/remote/api/MoviesApi.kt]: getMovieGenres

**Review by mariammnamgaladze — APPROVED** [review]

**Review by etozheluka — APPROVED** [review]

**Review by bchmsl — APPROVED** [review]

**Review by ninikokra — APPROVED** [review]

**Pull request merged**